File: Classes/CustomAutoCompleteCollection.vb
Option Strict On
Option Explicit On

Friend Class CustomAutoCompleteCollection
    Inherits AutoCompleteStringCollection

    Private ReadOnly _DataSource As New List(Of Dictionary(Of String, String))

    Public Sub New(ByVal dataSource As IDataReader, ByVal display As String)
        MyBase.New()

        Dim lIndex As Integer
        Dim lDataRow As Dictionary(Of String, String)

        While dataSource.Read
            lIndex = MyBase.Add(dataSource.Item(display).ToString)
            lDataRow = New Dictionary(Of String, String)

            For lItem As Integer = 0 To dataSource.FieldCount - 1
                lDataRow.Add(dataSource.GetName(lItem), CStr(dataSource.GetValue(lItem)))
            Next

            _DataSource.Insert(lIndex, lDataRow)
        End While
    End Sub

    Public Function GetValue(ByVal shortName As String, ByVal key As String) As String
        Dim lReturn As String

        If Not MyBase.Contains(shortName) Then Return String.Empty

        lReturn = _DataSource.Item(MyBase.IndexOf(shortName)).Item(key)

        Return lReturn
    End Function
End Class

File: QuickRun/Classes/CustomAutoCompleteCollection.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace QuickRun.Classes
{
    class CustomAutoCompleteCollection : AutoCompleteStringCollection
    {
        private readonly List<Dictionary<String, String>> _DataSource;

        public CustomAutoCompleteCollection(IDataReader dataSource, string display)
            : base()
        {
            int lIndex;
            Dictionary<string, string> lDataRow = new Dictionary<string, string>();

            while (dataSource.Read())
            {
                lIndex = base.Add(dataSource[display].ToString());
                lDataRow = new Dictionary<string, string>();

                for (int lItem = 0; lItem < dataSource.FieldCount; lItem++)
                {
                    lDataRow.Add(dataSource.GetName(lItem), dataSource.GetValue(lItem).ToString());
                }

                _DataSource.Insert(lIndex, lDataRow);
            }
        }

        public string GetValue(string shortName, string key)
        {
            string lReturn;

            if (!base.Contains(shortName))
            { return string.Empty; }

            lReturn = _DataSource[base.IndexOf(shortName)][key];

            return lReturn;
        }
    }
}
